daliuge-engine/dlg/io.py

```diff
@@ -151,39 +204,81 @@ def buffer(self) -> Union[memoryview, bytes, bytearray, pyarrow.Buffer]:
 
     @abstractmethod
     def _open(self, **kwargs):
+        """
+        Returns a reader or writer object depending on self._mode
+        """
         pass
 
     @abstractmethod
-    def _read(self, count, **kwargs):
+    def _read(self, count, **kwargs) -> bytes:
+        pass
+
+    #@abstractmethod
+    def _peek(self, count, **kwargs):
+        pass
+
+    #@abstractmethod
+    def _tell(self):
+        pass
+
+    #@abstractmethod
+    def _seek(self, offset, whence=io.SEEK_SET):
         pass
 
     @abstractmethod
     def _write(self, data, **kwargs) -> int:
         pass
 
     @abstractmethod
-    def _close(self, **kwargs):
+    def _close(self, **kwargs) -> None:
         pass
 
     @abstractmethod
     def _size(self, **kwargs) -> int:
         pass
 
 
+class AwaitOnce():
+    def __await__(self):
+        yield
+
+
+class DataIOAsyncIterator(AsyncIterator):
+    """
+    An async iterator for a drop using a dynamically
+    expanding buffer.
+    """
+    def __init__(self, memory_io: DataIO):
+        self._io = memory_io
+
+    async def __anext__(self):
+        #while not self._io.closed and self._io.tell() != self._io.size():
+        while self._io.tell() != self._io.size():
+            if self._io.peek(1):
+                return self._io.read(4096)
```

Comment:
this is a blocking read in a coroutine

Reply:
True. With a BytesIO object I'd expect this to be instant but if streaming was to support any storage-backed mediums the this would need something like an AsyncBufferedReader